### DISCUSSION TITLE
When I use contextvars, I unexpectedly found that FastAPI clears the content I set somewhere. Is this result normal?

### DISCUSSION THREAD

**Discussion by gaoKuoZqq** [fastapi/fastapi#13120] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import asyncio
import contextvars
import threading

import uvicorn
from fastapi import FastAPI

app = FastAPI()
contextvars_name = contextvars.ContextVar('contextvars_name', default="contextvars_name_default")


def print_task_id(mn: str):
    try:
        print(mn, "task_id", id(asyncio.current_task()))
    except Exception:
        print(mn, "task_id", "None")


def print_thread_id(flag: str):
    print(flag, "thread id:", threading.current_thread().ident)


def print_contextvars_name(flag: str):
    print(flag, "contextvars_name:", contextvars_name.get())


def print_all(flag: str):
    print_thread_id(flag)
    print_task_id(flag)
    print_contextvars_name(flag)


@app.get("/test1")
def test(cn: str | None = None):
    print_all("test1")
    if cn:
        contextvars_name.set(cn)
        print_all("test1")
    return "ok"

uvicorn.run(app)
```

### Description

With the same thread_id, I did not clear contextvars_name, but when I made the second request, the value of contextvars_name was not the one I set during the first request.
```
log:
INFO:     Started server process [27184]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     127.0.0.1:50183 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:50183 - "GET /openapi.json HTTP/1.1" 200 OK
test1 thread id: 2836
test1 task_id None
test1 contextvars_name: contextvars_name_default
test1 thread id: 2836
test1 task_id None
test1 contextvars_name: tt1
INFO:     127.0.0.1:50183 - "GET /test1?cn=tt1 HTTP/1.1" 200 OK
test1 thread id: 2836
test1 task_id None
test1 contextvars_name: contextvars_name_default
test1 thread id: 2836
test1 task_id None
test1 contextvars_name: tt1
INFO:     127.0.0.1:50183 - "GET /test1?cn=tt1 HTTP/1.1" 200 OK
```

### Operating System

Windows

### FastAPI Version

0.115.6

### Pydantic Version

2.10.4

### Python Version

3.11.11

### Additional Context

I thought that the contextvars_name printed on the second request should be tt1.

**Comment by sehraramiz** [comment, marked as answer]
Yes, ContextVar is a type of variable used to manage execution context for asynchronous operations. It was introduced to overcome the limitations of thread-local storage in asynchronous programming, as thread-local variables do not work well in async environments where multiple tasks share the same thread. For more information, see [PEP 567](https://peps.python.org/pep-0567/#rationale).

In the case of FastAPI, ContextVar can be used to set and access variables that are unique to each HTTP request. When a request starts, you can set a ContextVar in your route function or middleware, and every part of the code executed during that request (within the same async context) will have access to the variable. Meanwhile, if a second request comes in while the first is still in progress, the ContextVar for the second request is isolated and unaffected by the first.

I recommend experimenting with this concept using additional routes and adding some sleep functions to see how ContextVar behaves across concurrent requests.
also you can reset a ContextVar to it's previous value. [read this](https://docs.python.org/3.11/library/contextvars.html#contextvars.ContextVar.reset) 

Here's an example of an async route where you can observe task isolation using ContextVar:
```python

import asyncio
import contextvars
import threading
import time

from fastapi import FastAPI
import uvicorn

app = FastAPI()
contextvars_name = contextvars.ContextVar(
    "contextvars_name", default="contextvars_name_default"
)


def print_task_id(mn: str):
    try:
        print(mn, "task_id", id(asyncio.current_task()))
    except Exception:
        print(mn, "task_id", "None")


def print_thread_id(flag: str):
    print(flag, "thread id:", threading.current_thread().ident)


def print_contextvars_name(flag: str):
    print(flag, "contextvars_name:", contextvars_name.get())


def print_all(flag: str):
    print_thread_id(flag)
    print_task_id(flag)
    print_contextvars_name(flag)


@app.get("/test1")
def test(cn: str | None = None):
    print_all("test1")
    if cn:
        contextvars_name.set(cn)
        print_all("test1")
    time.sleep(5)
    return "ok"


@app.get("/test2")
async def test2(cn: str | None = None):
    print_all("test1")
    if cn:
        contextvars_name.set(cn)
        print_all("test1")
    await asyncio.sleep(5)
    return "ok"


uvicorn.run(app)

```
```console
curl "http://127.0.0.1:8000/test2?cn=var1"& curl "http://127.0.0.1:8000/test2?cn=var2"&
```

```
test2 thread id: 135637001094976
test2 task_id 135636939568896
test2 contextvars_name: contextvars_name_default
test2 thread id: 135637001094976
test2 task_id 135636939568896
test2 contextvars_name: var1
test2 thread id: 135637001094976
test2 task_id 135636939570048
test2 contextvars_name: contextvars_name_default
test2 thread id: 135637001094976
test2 task_id 135636939570048
test2 contextvars_name: var2
```

**Reply by gaoKuoZqq** [reply]
In my example, it can be seen that two requests are in the same thread. The second request should theoretically retrieve the value set by the first request, so I suspect that FastAPI has done something.

**Reply by Kludex** [reply]
No, the second request should not retrieve the value from the first request.

**Reply by gaoKuoZqq** [reply]
You are correct, but this is just a minimal example; we need to understand the reasons behind the disappearance of the values.

**Reply by Kludex** [reply]
Are you asking "why the thing that is supposed to happen is doing what is suppose to do?"

The contextvar shouldn't be propagated. Being in the same thread doesn't mean the value will not be reset.

**Reply by AndreasMietk** [reply]
**CAUTION**
`contextvars.ContextVar`s do not work as expected when used in conjunction with `BaseHTTPMiddleware` in Starlette/FastAPI

See: https://starlette.dev/middleware/#limitations

To get around this, rewrite your `BaseHTTPMiddleware`  as a _pure ASGI Middleware_!

Example:

**HTTPBaseMiddleware**
```python
class CustomHeaderMiddleware(BaseHTTPMiddleware):
    def __init__(self, app, header_value='Example'):
        super().__init__(app)
        self.header_value = header_value

    async def dispatch(self, request, call_next):
        response = await call_next(request)
        response.headers['Custom'] = self.header_value
        return response
```

VS

**ASGI Middleware**
```python
class CustomHeaderMiddleware:
    def __init__(self, app, header_value: str = 'Example'):
        self.app = app
        self.header_value = header_value.encode("latin-1")

    async def __call__(self, scope, receive, send):
        if scope["type"] != "http":
            await self.app(scope, receive, send)
            return

        async def send_with_custom_header(message):
            if message["type"] == "http.response.start":
                headers = list(message.get("headers", []))
                headers.append((b"custom", self.header_value))
                message["headers"] = headers
            
            await send(message)

        await self.app(scope, receive, send_with_custom_header)
        
 ```
